scripts/make_statistics.py

```diff
@@ -0,0 +1,539 @@
+    )
+
+    add_computational_stats(df_raw_osm_stats, snakemake)
+
+    return df_raw_osm_stats
+
+
+def collect_clean_osm_stats(rulename="clean_osm_data", metric_crs="EPSG:3857"):
+    """
+    Collect statistics on OSM data; used for clean OSM data.
+    """
+
+    snakemake = _mock_snakemake("clean_osm_data")
+
+    options_clean = dict(snakemake.output)
+    options_clean.pop("generators_csv")
+
+    df_clean_osm_stats = collect_osm_stats(
+        rulename, metric_crs=metric_crs, **options_clean
+    )
+
+    add_computational_stats(df_clean_osm_stats, snakemake)
+
+    return df_clean_osm_stats
+
+
+def collect_network_stats(network_rule, config):
+    """
+    Collect statistics on pypsa networks:
+    - installed capacity by carrier
+    - lines total length (accounting for parallel lines)
+    - lines total capacity
+    """
+
+    wildcards = {
+        k: str(config["scenario"][k][0]) for k in ["simpl", "clusters", "ll", "opts"]
+    }
+
+    snakemake = _mock_snakemake(network_rule, **wildcards)
+
+    network_path = (
+        snakemake.output["network"]
+        if "network" in snakemake.output.keys()
+        else snakemake.output[0]
+    )
+
+    def capacity_stats(df):
+        if df.empty:
+            return pd.Series(dtype=float)
+        else:
+            return df.groupby("carrier").p_nom.sum().astype(float)
+
+    if os.path.exists(network_path):
+        n = pypsa.Network(network_path)
+
+        lines_length = float((n.lines.length * n.lines.num_parallel).sum())
+
+        lines_capacity = float(n.lines.s_nom.sum())
+        buses_number = int(n.buses.shape[0])
+        buses_number_dc = (n.buses.carrier != "AC").sum()
+
+        network_stats = pd.DataFrame(
+            [[buses_number, buses_number_dc, lines_length, lines_capacity]],
+            columns=_multi_index_scen(
+                network_rule,
+                [
+                    "buses_number",
+                    "buses_number_non_ac",
+                    "lines_length",
+                    "lines_capacity",
+                ],
+            ),
+        )
+
+        gen_stats = pd.concat(
+            [capacity_stats(n.generators), capacity_stats(n.storage_units)], axis=0
+        )
+
+        if not gen_stats.empty:
+            df_gen_stats = gen_stats.to_frame().transpose().reset_index(drop=True)
+            df_gen_stats.columns = _multi_index_scen(network_rule, df_gen_stats.columns)
+            network_stats = pd.concat([network_stats, df_gen_stats], axis=1)
+
+        add_computational_stats(network_stats, snakemake)
+
+        return network_stats
+    else:
+        return pd.DataFrame()
+
+
+def collect_shape_stats(rulename="build_shapes", area_crs="ESRI:54009"):
+    """
+    Collect statistics on the shapes created by the workflow:
+    - area
+    - number of gadm shapes
+    - Percentage of shapes having country flag matching the gadm file
+    - total population
+    - total gdp
+    """
+
+    snakemake = _mock_snakemake(rulename)
```

Comment:
potentially not necessary, see comment few lines above

Reply:
same as above